File: leetcode/two_pointers/dart/trapping_rain_water.dart
import 'dart:math';

class Solution {
  int trap(List<int> height) {
    int left = 0,
        right = height.length - 1,
        leftMax = height[left],
        rightMax = height[right],
        result = 0;
    while (left < right) {
      if (leftMax < rightMax) {
        left++;
        leftMax = max(leftMax, height[left]);
        result += leftMax - height[left];
      } else {
        right--;
        rightMax = max(rightMax, height[right]);
        result += rightMax - height[right];
      }
    }
    return result;
  }
}

void main(List<String> args) {
  final height = <int>[0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];
  final object = Solution();
  print(object.trap(height));
}

File: leetcode/two_pointers/go/trapping_rain_water.go
package two_pointers

import "fmt"

func trap(height []int) int {
	var (
		left, right = 0, len(height) - 1
		leftMax     = height[left]
		rightMax    = height[right]
		result      = 0
	)
	for left < right {
		if leftMax < rightMax {
			left++
			leftMax = max(leftMax, height[left])
			result += leftMax - height[left]
		} else {
			right--
			rightMax = max(rightMax, height[right])
			result += rightMax - height[right]
		}
	}
	return result
}

func main() {
	height := []int{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}
	fmt.Println(trap(height))
}
